react-app/src/components/Extractor.js

```diff
@@ -0,0 +1,303 @@
+        type: 'file',
+        included: true,
+        key: 'filePath',
+        validExtractors: [],
+      },
+      {
+        url: props.formData.constructorArgs.url ? props.formData.constructorArgs.url : '',
+        label: 'URL',
+        type: 'url',
+        included: false,
+        key: 'url',
+        validExtractors: [],
+      },
+      {
+        clinicalSiteID: props.formData.constructorArgs.clinicalSiteID
+          ? props.formData.constructorArgs.clinicalSiteID
+          : '',
+        label: 'Clinical Site ID',
+        type: 'string',
+        included: false,
+        key: 'clinicalSiteID',
+        validExtractors: ['CSVClinicalTrialInformationExtractor'],
+      },
+      {
+        clinicalSiteSystem: props.formData.constructorArgs.clinicalSiteSystem
+          ? props.formData.constructorArgs.clinicalSiteSystem
+          : '',
+        label: 'Clinical Site System',
+        type: 'string',
+        included: false,
+        key: 'clinicalSiteSystem',
+        validExtractors: ['CSVClinicalTrialInformationExtractor'],
+      },
+      {
+        cancerType: props.formData.constructorArgs.cancerType ? props.formData.constructorArgs.cancerType : '',
+        label: 'Cancer Type',
+        type: 'string',
+        included: false,
+        key: 'cancerType',
+        validExtractors: ['CSVCancerDiseaseStatusExtractor'],
+      },
+      {
+        mask: props.formData.constructorArgs.mask ? props.formData.constructorArgs.mask : '',
+        label: 'Masked Fields',
+        type: 'dropdown',
+        included: false,
+        key: 'mask',
+        options: [`gender`, `mrn`, `name`, `address`, `birthDate`, `language`, `ethnicity`, `birthsex`, `race`],
+        validExtractors: ['CSVPatientExtractor'],
+      },
+    ];
+  }
+  const [args, setArgs] = useState(getArgValues());
+  const [argsJSX, setArgsJSX] = useState(
+    getArgValues()
+      .filter((arg) => arg.included === true)
+      .map((arg, i) => (
+        <p key={arg.key}>
+          This is a placeholder for arg {i}: {arg.label}
+        </p>
+      )),
+  );
+
+  function onExtractorLabelChange(e) {
+    props.onExtractorLabelChange(e.target.value, props.eventKey);
+    setExtractorLabel(e.target.value);
+  }
+
+  // FUNCTIONS FOR CONSTRUCTOR ARG MANAGEMENT
+
+  function getArgsJSX(tempArgs, updateArgData) {
+    return tempArgs
+      .filter((arg) => arg.included === true)
+      .map((arg, i) => {
+        switch (arg.type) {
+          case 'string':
+            return (
+              <Form.Group className="mb-3" controlId={arg.key} key={arg.key}>
+                <Form.Label>{arg.label}</Form.Label>
+                <div className="label-and-icon-container">
+                  <Form.Control
+                    type="text"
+                    value={arg[arg.key]}
+                    onChange={(e) => {
+                      const newArgs = [...tempArgs];
+                      newArgs.find((temp) => arg.key === temp.key)[arg.key] = e.target.value;
```

Comment:
I am curious what others think of this pattern, but there are a few places where you do a `find` and the directly manipulate the resulting object. While it doesn't make a big difference in the functionality of the code itself, I think it might make it easier to read this code in two lines like

```
const currentArg = newArgs.find((temp) => arg.key === temp.key);
currentArg[arg.key] = e.target.value;
```

There are a number of places with this pattern, so if we update this one, we should also update the others.

Reply:
I have replaced this code.

Reply:
Should I make a function to get the value for currentArg? This code repeats a lot, and that might simplify things.

Reply:
Yeah. It would be:
`const currentArg = getCurrentArg(arg.key);`